File: apple/packages/EnteComponents/Sources/EnteComponents/MenuRow.swift
import SwiftUI

public struct MenuRow<Leading: View>: View {
    @Environment(\.entePalette) private var palette
    private let title: String
    private let subtitle: String?
    private let leading: Leading?
    private let selected: Bool
    private let showsChevron: Bool
    private let action: () -> Void

    public init(
        _ title: String,
        subtitle: String? = nil,
        @ViewBuilder leading: () -> Leading,
        selected: Bool = false,
        showsChevron: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.subtitle = subtitle
        self.leading = leading()
        self.selected = selected
        self.showsChevron = showsChevron
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: EnteSpacing.md) {
                if let leading {
                    leading
                        .frame(width: 36, height: 36)
                        .foregroundStyle(palette.mutedText)
                        .accessibilityHidden(true)
                }
                VStack(alignment: .leading, spacing: EnteSpacing.xs) {
                    Text(title)
                        .font(EnteTypography.body)
                        .lineLimit(subtitle?.isEmpty == false ? 1 : 2)
                    if let subtitle, !subtitle.isEmpty {
                        Text(subtitle)
                            .font(EnteTypography.mini)
                            .foregroundStyle(palette.mutedText)
                            .lineLimit(2)
                    }
                }
                Spacer(minLength: 0)
                if selected || showsChevron {
                    HStack(spacing: EnteSpacing.xs) {
                        if selected {
                            Glyph.check.image.resizable().scaledToFit()
                                .frame(width: 24, height: 24)
                                .foregroundStyle(palette.primary)
                        }
                        if showsChevron {
                            Glyph.chevron.image.resizable().scaledToFit()
                                .frame(width: 24, height: 24)
                                .flipsForRightToLeftLayoutDirection(true)
                                .foregroundStyle(palette.mutedText)
                        }
                    }
                    .frame(minWidth: 36, minHeight: 36)
                    .accessibilityHidden(true)
                }
            }
            .padding(.leading, leading == nil ? EnteSpacing.lg : EnteSpacing.md)
            .padding(.trailing, EnteSpacing.md)
            .padding(.vertical, 9)
            .frame(minHeight: 58)
            .contentShape(Rectangle())
        }
        .buttonStyle(MenuRowStyle())
        .accessibilityAddTraits(selected ? .isSelected : [])
    }
}

public extension MenuRow where Leading == EmptyView {
    init(
        _ title: String,
        subtitle: String? = nil,
        selected: Bool = false,
        showsChevron: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.subtitle = subtitle
        leading = nil
        self.selected = selected
        self.showsChevron = showsChevron
        self.action = action
    }
}

private struct MenuRowStyle: ButtonStyle {
    @Environment(\.entePalette) private var palette
    @Environment(\.isEnabled) private var enabled

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .foregroundStyle(enabled ? palette.text : palette.disabledText)
            .background(configuration.isPressed ? palette.fillDarker : palette.surface)
            .animation(.easeOut(duration: EnteMotion.quick), value: configuration.isPressed)
    }
}

File: android/packages/components/src/main/kotlin/io/ente/components/MenuRow.kt
package io.ente.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

@Composable
public fun MenuRow(
    title: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    selected: Boolean? = null,
    showsChevron: Boolean = false,
    enabled: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val background by
        animateColorAsState(
            targetValue = if (pressed && enabled) palette.fillDarker else palette.surface,
            animationSpec = tween(EnteMotion.quick),
            label = "menuRowBackground",
        )
    val hasSubtitle = !subtitle.isNullOrEmpty()

    Row(
        modifier
            .fillMaxWidth()
            .heightIn(min = 58.dp)
            .background(background)
            .clickable(
                enabled = enabled,
                role = Role.Button,
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick,
            )
            .semantics { selected?.let { this.selected = it } }
            .padding(
                start = if (leading == null) EnteSpacing.lg else EnteSpacing.md,
                end = EnteSpacing.md,
                top = 9.dp,
                bottom = 9.dp,
            ),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(EnteSpacing.md),
    ) {
        leading?.let {
            CompositionLocalProvider(LocalContentColor provides palette.mutedText) {
                Box(Modifier.size(36.dp), contentAlignment = Alignment.Center) { it() }
            }
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(EnteSpacing.xs)) {
            Text(
                title,
                maxLines = if (hasSubtitle) 1 else 2,
                overflow = TextOverflow.Ellipsis,
                style = EnteTypography.body,
                color = if (enabled) palette.text else palette.disabledText,
            )
            if (!subtitle.isNullOrEmpty()) {
                Text(
                    subtitle,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    style = EnteTypography.mini,
                    color = palette.mutedText,
                )
            }
        }
        if (selected == true || showsChevron) {
            Row(
                Modifier.defaultMinSize(minWidth = 36.dp, minHeight = 36.dp),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (selected == true) {
                    Icon(Icons.Default.Check, null, Modifier.size(24.dp), tint = palette.primary)
                }
                if (showsChevron) {
                    if (selected == true) Spacer(Modifier.width(EnteSpacing.xs))
                    Icon(
                        Icons.AutoMirrored.Filled.KeyboardArrowRight,
                        null,
                        Modifier.size(24.dp),
                        tint = palette.mutedText,
                    )
                }
            }
        }
    }
}
